File: app/src/main/java/suzp1984/github/io/exapidemo/app/system/RuntimeMemoryActivity.java
package suzp1984.github.io.exapidemo.app.system;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.Button;
import android.widget.TextView;

import java.util.LinkedList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import suzp1984.github.io.exapidemo.R;

/**
 * Created by suzhenxi on 10/27/2016.
 */

public class RuntimeMemoryActivity extends AppCompatActivity {

    @BindView(R.id.runtime_check)
    Button mRuntimeCheckBtn;

    @BindView(R.id.allocate_20_memory)
    Button mAllocate20MBtn;

    @BindView(R.id.allocate_all_memory)
    Button mAllocateAllMemory;

    @BindView(R.id.memory_txt)
    TextView mMemoryTxt;

    final List<byte[]> mArrays = new LinkedList<>();

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_runtime_memory);

        ButterKnife.bind(this);
    }

    @OnClick(R.id.runtime_check)
    public void runtimeCheck() {
        final Runtime runtime = Runtime.getRuntime();
        long free = runtime.freeMemory();
        long max = runtime.maxMemory();
        long total = runtime.totalMemory();

        mMemoryTxt.setText("free: " + String.valueOf(free) + "K = " + free/1024L/1024L + "M; "
                + "total: " + String.valueOf(total) + "K = " + total/1024L/1024L + "M;"
                + "whole: " + String.valueOf(max) + "K = " + max/1024L/1024L + "M;");
    }

    @OnClick(R.id.allocate_all_memory)
    public void allocateAllMemory() {
        final Runtime runtime = Runtime.getRuntime();
        long remain = runtime.maxMemory() - runtime.totalMemory();

        byte[] block = new byte[(int)remain];

        mArrays.add(block);

        runtimeCheck();
    }

    @OnClick(R.id.allocate_20_memory)
    public void allocate20Memory() {

        byte[] twn = new byte[(int) (1024L*1024L*20)];

        mArrays.add(twn);

        runtimeCheck();
    }
}

File: app/src/main/kotlin/suzp1984/github/io/exapidemo/app/system/RuntimeMemoryActivity.kt
package suzp1984.github.io.exapidemo.app.system

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.widget.Button
import android.widget.TextView

import java.util.LinkedList

import butterknife.BindView
import butterknife.ButterKnife
import butterknife.OnClick
import suzp1984.github.io.exapidemo.R

/**
 * Created by suzhenxi on 10/27/2016.
 */

class RuntimeMemoryActivity : AppCompatActivity() {

    @BindView(R.id.runtime_check)
    lateinit var mRuntimeCheckBtn: Button

    @BindView(R.id.allocate_20_memory)
    lateinit var mAllocate20MBtn: Button

    @BindView(R.id.allocate_all_memory)
    lateinit var mAllocateAllMemory: Button

    @BindView(R.id.memory_txt)
    lateinit var mMemoryTxt: TextView

    internal val mArrays: MutableList<ByteArray> = LinkedList()

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_runtime_memory)

        ButterKnife.bind(this)
    }

    @OnClick(R.id.runtime_check)
    fun runtimeCheck() {
        val runtime = Runtime.getRuntime()
        val free = runtime.freeMemory()
        val max = runtime.maxMemory()
        val total = runtime.totalMemory()

        mMemoryTxt.text = "free: ${free}K = ${free / 1024L / 1024L}M; " +
                    "total: ${total}K = ${total / 1024L / 1024L}M;" +
                    "whole: ${max}K = ${max / 1024L / 1024L}M;"
    }

    @OnClick(R.id.allocate_all_memory)
    fun allocateAllMemory() {
        val runtime = Runtime.getRuntime()
        val remain = runtime.maxMemory() - runtime.totalMemory()

        try {
            val block = ByteArray(remain.toInt())

            mArrays.add(block)

            runtimeCheck()
        } catch (e: OutOfMemoryError) {
            e.printStackTrace()
        }
    }

    @OnClick(R.id.allocate_20_memory)
    fun allocate20Memory() {

        try {
            val twn = ByteArray((1024L * 1024L * 20).toInt())

            mArrays.add(twn)

            runtimeCheck()
        } catch (e: OutOfMemoryError) {
            e.printStackTrace()
        }
    }
}
